crates/sui-core/src/execution_cache/writeback_cache.rs

```diff
@@ -0,0 +1,1158 @@
+                        // that the object does not exist anywhere
+                        return CacheResult::NegativeHit;
+                    }
+                }
+            };
+        }
+
+        if let Some(objects) = self.dirty.objects.get(object_id) {
+            check_cache_entry!(objects);
+        }
+
+        if let Some(objects) = self.cached.object_cache.get(object_id) {
+            let objects = objects.lock();
+            check_cache_entry!(objects);
+        }
+
+        CacheResult::Miss
+    }
+
+    fn get_object_by_key_cache_only(
+        &self,
+        object_id: &ObjectID,
+        version: SequenceNumber,
+    ) -> CacheResult<Object> {
+        match self.get_object_entry_by_key_cache_only(object_id, version) {
+            CacheResult::Hit(entry) => match entry {
+                ObjectEntry::Object(object) => CacheResult::Hit(object),
+                ObjectEntry::Deleted | ObjectEntry::Wrapped => CacheResult::NegativeHit,
+            },
+            CacheResult::Miss => CacheResult::Miss,
+            CacheResult::NegativeHit => CacheResult::NegativeHit,
+        }
+    }
+
+    fn get_object_entry_by_id_cache_only(
+        &self,
+        object_id: &ObjectID,
+    ) -> CacheResult<(SequenceNumber, ObjectEntry)> {
+        macro_rules! check_cache_entry {
+            ($objects: expr) => {
+                if let Some((version, entry)) = $objects.get_last() {
+                    return CacheResult::Hit((*version, entry.clone()));
+                } else {
+                    return CacheResult::Miss;
+                }
+            };
+        }
+
+        if let Some(objects) = self.dirty.objects.get(object_id) {
+            check_cache_entry!(objects);
+        }
+
+        if let Some(objects) = self.cached.object_cache.get(object_id) {
+            let objects = objects.lock();
+            check_cache_entry!(objects);
+        }
+
+        CacheResult::Miss
+    }
+
+    fn get_object_by_id_cache_only(
+        &self,
+        object_id: &ObjectID,
+    ) -> CacheResult<(SequenceNumber, Object)> {
+        match self.get_object_entry_by_id_cache_only(object_id) {
+            CacheResult::Hit((version, entry)) => match entry {
+                ObjectEntry::Object(object) => CacheResult::Hit((version, object)),
+                ObjectEntry::Deleted | ObjectEntry::Wrapped => CacheResult::NegativeHit,
+            },
+            CacheResult::NegativeHit => CacheResult::NegativeHit,
+            CacheResult::Miss => CacheResult::Miss,
+        }
+    }
+
+    // Commits dirty data for the given TransactionDigest to the db.
+    async fn commit_transaction_outputs(
+        &self,
+        epoch: EpochId,
+        digest: TransactionDigest,
+    ) -> SuiResult {
+        let Some((_, outputs)) = self.dirty.pending_transaction_writes.remove(&digest) else {
+            return Err(SuiError::TransactionNotFound { digest });
```

Comment:
Could you comment explain when this might happen and why it's not a panic?

Reply:
If we allow local execution on FN to call commit_transaction, then it might race with CheckpointExecutor. I think we decided that we can't do that anyway, but it feels like the panic belongs in the caller